ptn_utils/logger/logger.py

```diff
@@ -0,0 +1,146 @@
+
+    def filter_function(record: dict) -> bool:
+        record_logger_name = record["extra"].get("logger_name", []).split(".")
+        for logger_name in LOG_SINKS:
+            if logger_name == "_default":
+                continue
+            logger_name_list = logger_name.split(".")
+            if len(record_logger_name) >= len(logger_name_list):
+                if record_logger_name[: len(logger_name_list)] == logger_name_list:
+                    return False
+        return True
+
+    sink_id = logger.add(
+        stdout,
+        level=level,
+        filter=filter_function,
+    )
+    LOG_SINKS["_default"] = sink_id
+
+
+def create_logger_sink(logger_name: str, level: str) -> None:
+    if logger_name in LOG_SINKS:
+        logger.remove(LOG_SINKS[logger_name])
+
+    def filter_function(record: dict) -> bool:
+        record_logger_name = record["extra"].get("logger_name", []).split(".")
+        logger_name_list = logger_name.split(".")
+        if len(record_logger_name) >= len(logger_name_list):
+            return record_logger_name[: len(logger_name_list)] == logger_name_list
+        return False
+
+    sink_id = logger.add(
+        stdout,
+        level=level,
+        filter=filter_function,
+    )
+    LOG_SINKS[logger_name] = sink_id
+
+
+def setup_logging() -> None:
+    logger.info("Setting up logging configuration.")
+
+    # Send all logging through loguru
+    log_handler = InterceptHandler()
+    logging.root.handlers = [log_handler]
+    logging.root.setLevel(logging.DEBUG)
+
+    # Set default logging level from environment variable or INFO
+    loglevel_input = os.getenv("PTN_LOG_LEVEL", "INFO")
+    logger.remove()
+    create_default_logger_sink(loglevel_input)
+
+    logger.info(f"Logging level set to {loglevel_input}.")
+
+
+class LogLevels(Enum):
+    Critical = "CRITICAL"
+    Error = "ERROR"
+    Warning = "WARNING"
+    Info = "INFO"
+    Debug = "DEBUG"
+
+
+async def set_logging_level_autocomplete(
+    interaction: Interaction,
+    current: str,
+) -> List[app_commands.Choice[str]]:
+    all_loggers = [logging.getLogger(name).name for name in logging.root.manager.loggerDict]
+    if "." in current.lower():
+        pass  # we've already specified a package. Shouldn't have more than 25 loggers in one package, _right_?
+    else:
+        all_loggers = [x for x in all_loggers if "." not in x]  # yeet subpackages if we haven't selected a package yet
+
+    if len(all_loggers) > 25:
+        """
+        Safety valve in case one or both of the assumptions above fail:
+            - >25 top-level packages or 
+            - >25 loggers for a given top-level package
+        
+        Generate a warning and move on. Log the full list in debug if we care to check it out later
+        """
+
+        logger.warning("Autocomplete returned more options than Discord can handle. Truncating to 25")
+        logger.debug(all_loggers)
+        all_loggers = all_loggers[1:25]
+
```

Comment:
```suggestion
    all_loggers = all_loggers[:25]
```

I can't see a reason to discard the first one.
Also moved truncate out of the condition for extra safety